server.go

```diff
@@ -545,41 +541,32 @@ func (p *sshFxpFsetstatPacket) respond(svr *Server) responsePacket {
 		return statusFromError(p.ID, EBADF)
 	}
 
-	// additional unmarshalling is required for each possibility here
-	b := p.Attrs.([]byte)
+	path := f.Name()
+
+	debug("fsetstat name %q", path)
+
+	fs := p.unmarshalFileStat(p.Flags)
+	
 	var err error
 
-	debug("fsetstat name \"%s\"", f.Name())
 	if (p.Flags & sshFileXferAttrSize) != 0 {
-		var size uint64
-		if size, b, err = unmarshalUint64Safe(b); err == nil {
-			err = f.Truncate(int64(size))
+		if err == nil {
+			err = f.Truncate(int64(fs.Size))
 		}
 	}
 	if (p.Flags & sshFileXferAttrPermissions) != 0 {
-		var mode uint32
-		if mode, b, err = unmarshalUint32Safe(b); err == nil {
-			err = f.Chmod(os.FileMode(mode))
+		if err == nil {
+			err = f.Chmod(fs.FileMode())
 		}
 	}
 	if (p.Flags & sshFileXferAttrACmodTime) != 0 {
-		var atime uint32
-		var mtime uint32
-		if atime, b, err = unmarshalUint32Safe(b); err != nil {
-		} else if mtime, b, err = unmarshalUint32Safe(b); err != nil {
-		} else {
-			atimeT := time.Unix(int64(atime), 0)
-			mtimeT := time.Unix(int64(mtime), 0)
-			err = os.Chtimes(f.Name(), atimeT, mtimeT)
+		if err == nil {
+			err = os.Chtimes(path, fs.AccessTime(), fs.ModTime())
```

Comment:
The only operation here that doesn’t have an “on a handle` operation.

🤔 This should probably actually be last as well. I don’t know if changing the UID/GID below would reset the the `mtime` or the `atime`, right?